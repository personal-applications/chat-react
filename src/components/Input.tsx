import React, { forwardRef } from "react";
import { UseFormRegisterReturn } from "react-hook-form";
import { If, Then } from "react-if";
import { twMerge } from "tailwind-merge";

export type InputProps = Partial<UseFormRegisterReturn> &
  Pick<React.InputHTMLAttributes<HTMLInputElement>, "placeholder" | "type"> & {
    error?: string;
    postIcon?: React.ReactElement;
  };

const Input = forwardRef(function Input(
  { error, type = "text", postIcon, ...props }: InputProps,
  ref: React.Ref<HTMLInputElement>,
) {
  return (
    <div className="flex w-full flex-col">
      <label
        className={twMerge(
          "input input-bordered flex items-center gap-2",
          error && "input-error",
        )}
      >
        <input type={type} className="grow" {...props} ref={ref} />
        <If condition={React.isValidElement(postIcon)}>
          <Then>{postIcon}</Then>
        </If>
      </label>

      <If condition={error}>
        <Then>
          <span className="pt-3 text-sm text-error">{error}</span>
        </Then>
      </If>
    </div>
  );
});

export default Input;
