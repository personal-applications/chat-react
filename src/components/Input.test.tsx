import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Input from "./Input";
import React from "react";

test("displays error message correctly", () => {
  render(<Input placeholder="Enter your name" error="Invalid input" />);

  expect(screen.getByText("Invalid input")).toBeInTheDocument();
});

test("forwards ref correctly", () => {
  const ref = React.createRef<HTMLInputElement>();
  render(<Input placeholder="Enter your name" ref={ref} />);

  expect(ref.current).toBeInstanceOf(HTMLInputElement);
});

test("handles input correctly", async () => {
  const handleChange = vi.fn();
  render(<Input placeholder="Enter your name" onChange={handleChange} />);

  const input = screen.getByPlaceholderText(
    "Enter your name",
  ) as HTMLInputElement;
  await userEvent.type(input, "John Doe");

  expect(handleChange).toHaveBeenCalledTimes(8); // One event per character
  expect(input.value).toBe("John Doe");
});
