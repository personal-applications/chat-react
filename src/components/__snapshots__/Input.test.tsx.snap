// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`displays postIcon correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex w-full flex-col"
      >
        <label
          class="input input-bordered flex items-center gap-2"
        >
          <input
            class="grow"
            placeholder="Enter your name"
            type="text"
          />
          <svg
            class="w-4 h-4 opacity-70"
            fill="currentColor"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M9.965 11.026a5 5 0 1 1 1.06-1.06l2.755 2.754a.75.75 0 1 1-1.06 1.06l-2.755-2.754ZM10.5 7a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Z"
              fill-rule="evenodd"
            />
          </svg>
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex w-full flex-col"
    >
      <label
        class="input input-bordered flex items-center gap-2"
      >
        <input
          class="grow"
          placeholder="Enter your name"
          type="text"
        />
        <svg
          class="w-4 h-4 opacity-70"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M9.965 11.026a5 5 0 1 1 1.06-1.06l2.755 2.754a.75.75 0 1 1-1.06 1.06l-2.755-2.754ZM10.5 7a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Z"
            fill-rule="evenodd"
          />
        </svg>
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
